### PR TITLE
Feature: better lambda telemetry

### PR DESCRIPTION
**Pull request by lbeschastny**
This PR significantly improves telemetry traces reporting from rust lambdas:

 * Traces are now sent reliably. Previously traces were often lost.
 * Lambda now always reports a root span with all required metadata.

### CODE REVIEW

File: src/middleware/lambda.rs [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    type Future = InstrumentedFuture<Instrumented<S::Future>, TracerProvider>;
+
+    fn poll_ready(&mut self, cx: &mut TaskContext<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: LambdaInvocation) -> Self::Future {
+        let span = tracing::info_span!(
+            "Lambda function invocation",
+            { semconv::FAAS_TRIGGER } = "http",
```
[the remark this answers is not in the document]
Comment (lbeschastny) [reply src/middleware/lambda.rs:60]: Also, we could extract telemetry context from `req.context.xray_trace_id` here.

File: src/middleware/lambda.rs [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+            .context
+            .xray_trace_id
+            .as_deref()
+            .and_then(span_context_from_str)
+            .map(|span_context| {
+                Context::map_current(|cx| cx.with_remote_span_context(span_context))
+            });
+        if let Some(cx) = context {
+            span.set_parent(cx);
+        }
```
Comment (lbeschastny) [inline src/middleware/lambda.rs:81]: Do we care about preserving X-Ray context?

File: src/middleware/lambda.rs [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+    fn call(&mut self, req: LambdaInvocation) -> Self::Future {
+        let span = tracing::trace_span!(
+            target: TRACING_TARGET,
+            "Lambda function invocation",
+            // TODO: set correct otel.kind and faas.trigger
+            // see https://opentelemetry.io/docs/specs/semconv/faas/aws-lambda/
+            "otel.kind" = ?SpanKind::Server,
+            "otel.name" = req.context.env_config.function_name,
+            { semconv::FAAS_TRIGGER } = "other",
```
Comment (lbeschastny) [inline src/middleware/lambda.rs:71]: We should follow [OpenTelemetry Semantic Conventions for AWS Lambda](https://opentelemetry.io/docs/specs/semconv/faas/aws-lambda/) here

File: src/middleware/lambda.rs (hunk 2 above)
Comment (lbeschastny) [reply src/middleware/lambda.rs:87]: I'll remove this for now

File: src/middleware/lambda.rs (hunk 3 above)
Comment (lbeschastny) [reply src/middleware/lambda.rs:67]: I'll do it later in a separate PR

**Review by Travispersson — APPROVED** [review]

**Pull request merged**